src/EffectNode.cpp

```diff
@@ -15,105 +15,91 @@
-    o.insert("intensity", m_intensity);
+    o.insert("file", d()->m_file);
+    o.insert("intensity", d()->m_intensity);
     return o;
 }
 
-void EffectNode::onInitialized() {
+void EffectNode::chainsEdited(QList<Chain> added, QList<Chain> removed) {
+    for (auto chain : added) {
+        auto result = QMetaObject::invokeMethod(d()->m_openGLWorker.data(), "addNewState", Q_ARG(Chain, chain));
+        Q_ASSERT(result);
+    }
     {
-        QMutexLocker locker(&m_stateLock);
-        for (auto key : m_renderStates.keys()) {
-            auto state = QSharedPointer<EffectNodeRenderState>::create();
-            m_openGLWorker->prepareState(state);
-            m_renderStates[key].swap(state);
+        QMutexLocker locker(&d()->m_stateLock);
+        for (int i=0; i<removed.count(); i++) {
+            d()->m_renderStates.remove(removed.at(i));
         }
     }
-    m_ready = true;
-}
-
-void EffectNode::chainsEdited(QList<QSharedPointer<Chain>> added, QList<QSharedPointer<Chain>> removed) {
-    QMutexLocker locker(&m_stateLock);
-    for (int i=0; i<added.count(); i++) {
-        auto state = QSharedPointer<EffectNodeRenderState>::create();
-        m_openGLWorker->prepareState(state);
-        m_renderStates.insert(added.at(i), state);
-    }
-    for (int i=0; i<removed.count(); i++) {
-        m_renderStates.remove(removed.at(i));
-    }
 }
 
-// Paint never needs to take the stateLock
-// because it already has a copy of the object
-// and this copy will never be modified
-// out from under it, unlike the parent
-// from which it was created
-GLuint EffectNode::paint(QSharedPointer<Chain> chain, QVector<GLuint> inputTextures) {
-    // Hitting this assert means
-    // that you failed to make a copy
-    // of the VideoNode
-    // before rendering in a different thread
-    Q_ASSERT(QThread::currentThread() == thread());
-
+GLuint EffectNode::paint(Chain chain, QVector<GLuint> inputTextures) {
     GLuint outTexture = 0;
 
-    if (!m_ready) {
-        //qDebug() << this << "is not ready";
-        return inputTextures.at(0); // Pass-through
-    }
-    //qDebug() << "Looking up" << chain << "in" << m_renderStates << "of" << this;
-    if (!m_renderStates.contains(chain)) { // This check doesn't seem to work
-        qDebug() << this << "does not have chain" << chain;
-        return inputTextures.at(0);
+    QSharedPointer<EffectNodeRenderState> renderState;
+    int inputCount;
+    auto time = context()->timebase()->beat();
+    auto wallTime = context()->timebase()->wallTime();
+    qreal step;
+    qreal intensityIntegral;
+    qreal frequency;
+    {
+        QMutexLocker locker(&d()->m_stateLock);
+        if (!d()->m_ready) {
+            //qDebug() << this << "is not ready";
+            return inputTextures.at(0); // Pass-through
+        }
+        //qDebug() << "Looking up" << chain << "in" << d()->m_renderStates << "of" << this;
+        if (!d()->m_renderStates.contains(chain)) { // This check doesn't seem to work
```

Comment:
What does this comment mean, it sounds bad

Reply:
This comment existed in the original code, I didn't change it. I can delete it?

Reply:
Confirmed, this check does in fact work